src/we-blog-log.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "WE-BLOG-LOG 1"
.TH WE-BLOG-LOG 1 "2012-07-10" "Version 0.9" "We-Blog Documentation"
.SH "NAME"
we\-blog\-log \- displays the We\-Blog repository log
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBwe-blog-log\fR [\fB\-cqrsCV\fR] [\fB\-b\fR \fIdirectory\fR] [\fB\-n\fR \fInumber\fR]
.PP
\&\fBwe-blog-log\fR \fB\-h\fR|\fB\-v\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBwe-blog-log\fR displays the content of the We-Blog repository log.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-b\fR \fIdirectory\fR, \fB\-\-blogdir\fR \fIdirectory\fR" 4
.IX Item "-b directory, --blogdir directory"
Allows you to specify a \fIdirectory\fR in which the We-Blog repository
is placed. The default option is a current working directory.
.IP "\fB\-n\fR \fInumber\fR, \fB\-\-number\fR \fInumber\fR" 4
.IX Item "-n number, --number number"
Allows you to specify a \fInumber\fR of log entries to be listed.
.IP "\fB\-s\fR, \fB\-\-short\fR" 4
.IX Item "-s, --short"
Tells \fBwe-blog-log\fR to display each log entry on a single line.
.IP "\fB\-r\fR, \fB\-\-reverse\fR" 4
.IX Item "-r, --reverse"
Tells \fBwe-blog-log\fR to display log entries in reverse order.
.IP "\fB\-c\fR, \fB\-\-color\fR" 4
.IX Item "-c, --color"
Enables colored output. When supplied, this option overrides the relevant
configuration option.
.IP "\fB\-C\fR, \fB\-\-no\-color\fR" 4
.IX Item "-C, --no-color"
Disables colored output. When supplied, this option overrides the relevant
configuration option.
.IP "\fB\-q\fR, \fB\-\-quiet\fR" 4
.IX Item "-q, --quiet"
Disables displaying of unnecessary messages.
.IP "\fB\-V\fR, \fB\-\-verbose\fR" 4
.IX Item "-V, --verbose"
Enables displaying of all messages. This is the default option.
.IP "\fB\-h\fR, \fB\-\-help\fR" 4
.IX Item "-h, --help"
Displays usage information and exits.
.IP "\fB\-v\fR, \fB\-\-version\fR" 4
.IX Item "-v, --version"
Displays version information and exits.
.SH "FILES"
.IX Header "FILES"
.IP "\fI.we\-blog/log\fR" 4
.IX Item ".we-blog/log"
A file containing the repository log.
.SH "EXAMPLE USAGE"
.IX Header "EXAMPLE USAGE"
List the whole repository history:
.PP
.Vb 2
\&        $ we-blog-log
\&        Date: Sun Jul 25 16:48:22 2010
.Ve
.PP
.Vb 1
\&                        Edited the page with ID 5.
.Ve
.PP
.Vb 1
\&        Date: Tue Jul  6 18:54:59 2010
.Ve
.PP
.Vb 1
\&                        Edited the page with ID 5.
.Ve
.PP
.Vb 1
\&        etc.
.Ve
.PP
List the whole repository history in reverse order:
.PP
.Vb 2
\&        $ we-blog-log -r
\&        Date: Tue Feb 10 00:40:16 2009
.Ve
.PP
.Vb 1
\&                        Created/recovered a We-Blog repository.
.Ve
.PP
.Vb 1
\&        Date: Tue Feb 10 01:06:44 2009
.Ve
.PP
.Vb 1
\&                        Added the page with ID 1.
.Ve
.PP
.Vb 1
\&        etc.
.Ve
.PP
Display the very first log entry on a single line:
.PP
.Vb 2
\&        $ we-blog-log -rs -n 1
\&        Tue Feb 10 00:40:16 2009 - Created/recovered a We-Blog repository.
.Ve
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBwe-blog-init\fR(1), \fBwe-blog-config\fR(1)
.SH "BUGS"
.IX Header "BUGS"
To report a bug or to send a patch, please, add a new issue to the bug
tracker at <http://code.google.com/p/we\-blog/issues/>, or visit the
discussion group at <https://groups.google.com/d/forum/tonk\-we\-blog>.
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2008\-2011 Jaromir Hradilek / 2011\-2012 Ton Kersten
.PP
This program is free software; see the source for copying conditions. It is
distributed in the hope that it will be useful, but \s-1WITHOUT\s0 \s-1ANY\s0 \s-1WARRANTY\s0;
without even the implied warranty of \s-1MERCHANTABILITY\s0 or \s-1FITNESS\s0 \s-1FOR\s0 A
\&\s-1PARTICULAR\s0 \s-1PURPOSE\s0.


ChangeLog
2012-07-10  Ton Kersten <we-blog@tonkersten.com>
  * Declared 0.8 stable
  * Switched to branch 0.9 for further development

2012-07-09  Ton Kersten <we-blog@tonkersten.com>
  * Changed a lot of things, but they all boil down to
    bugfixes and code reshuffeling

2011-10-07  Ton Kersten <we-blog@tonkersten.com>
  * Added 'alt' text for pages to show on mouse-over.

2011-09-21  Ton Kersten <we-blog@tonkersten.com>
  * And switched to branch 0.8 for further development

2011-09-21  Ton Kersten <we-blog@tonkersten.com>
  * Increased we-blog to version 0.7
  * Added 'smiley' converter
  * Published the project on GitHub

2011-08-15  Ton Kersten <we-blog@tonkersten.com>
  * Forked BlazeBlogger 1.2.0 to we-blog 0.5
    This because development of Blaze seems to be stalled

2011-02-18  Jaromir Hradilek <jhradilek@gmail.com>

  * INSTALL, Makefile, README, src/we-blog-add.pl, src/we-blog-config.pl,
  src/we-blog-edit.pl, src/we-blog-init.pl, src/we-blog-list.pl, src/we-blog-log.pl,
  src/we-blog-make.pl, src/we-blog-remove.pl, unix/we-blog.sh, unix/man/man1/we-blog.1,
  windows/we-blog.bat: Increased the version number.

  * unix/man/man1/we-blog.1: Updated the manual page for the command wrapper.

# $Id: ChangeLog 4 2012-07-10 09:34:11 tonk $


src/we-blog-log.pl
#!/usr/bin/env perl
# vi: set sw=4 ts=4 ai:
# $Id: we-blog-log.pl 2 2011-09-21 15:34:33 tonk $

# we-blog-log - displays the We-Blog repository log
# Copyright (c) 2011-2012 Ton Kersten
# Copyright (c) 2009-2011 Jaromir Hradilek

# This program is  free software:  you can redistribute it and/or modify it
# under  the terms  of the  GNU General Public License  as published by the
# Free Software Foundation, version 3 of the License.
#
# This program  is  distributed  in the hope  that it will  be useful,  but
# WITHOUT  ANY WARRANTY;  without  even the implied  warranty of MERCHANTA-
# BILITY  or  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
# License for more details.
#
# You should have received a copy of the  GNU General Public License  along
# with this program. If not, see <http://www.gnu.org/licenses/>.

use strict;
use warnings;
use File::Basename;
use File::Spec::Functions;
use Getopt::Long;
use Term::ANSIColor;
use Text::Wrap;

# Set the library path and use our own module
use lib dirname($0);
use We;

# Display usage information:
sub display_help {
	# Display the usage:
	print << "END_HELP";
Usage: $NAME [-cqrsCV] [-b DIRECTORY] [-n NUMBER]
			 $NAME -h|-v

	-b, --blogdir DIRECTORY     specify a directory in which the We-Blog
	                            repository is placed
	-n, --number NUMBER         specify a number of log entries to be listed
	-s, --short                 display each log entry on a single line
	-r, --reverse               display log entries in reverse order
	-c, --color                 enable colored output
	-C, --no-color              disable colored output
	-q, --quiet                 do not display unnecessary messages
	-V, --verbose               display all messages
	-h, --help                  display this help and exit
	-v, --version               display version information and exit
END_HELP

	# Return success:
	return 1;
}

# Display a log entry:
sub display_record {
	my $record = shift || die 'Missing argument';

	# Check whether to use compact listing:
	unless ($compact) {
		# Decompose the record:
		my ($date, $message) = split(/\s+-\s+/, $record, 2);

		# Check whether colors are enabled:
		unless ($coloured) {
			# Display plain record header:
			print "Date: $date\n\n";
		}
		else {
			# Display colored record header:
			print colored ("Date: $date", 'yellow');
			print "\n\n";
		}

		# Display the record body:
		print wrap('    ', '    ', $message);
		print "\n";
	}
	else {
		# Display the short record:
		print $record;
	}

	# Return success:
	return 1;
}

# Display log entries:
sub display_log {
	# Initialize required variables:
	my @lines = ();
	my $count = 0;

	# Prepare the file name:
	my $file  = catfile($blogdir, $weblog, 'log');

	# Open the log file for reading:
	open(LOG, "$file") or return 0;

	# Process each entŕy:
	while (my $record = <LOG>) {
		# Check whether to use reverse order:
		if ($reverse) {
			# Display the log entry immediately:
			display_record($record);

			# Check whether the limited number of displayed entries is requested:
			if ($number > 0) {
				# Increase the displayed entries counter:
				$count++;

				# End loop when the counter reaches the limit:
				last if $count == $number;
			}
		}
		else {
			# Prepend the log entry to the list of records to be displayed later:
			unshift(@lines, $record);
		}
	}

	# Close the file:
	close(LOG);

	# Unless the reverse order was requested, and therefore records have been
	# already displayed, display them now:
	unless ($reverse) {
		# Process each log entry:
		foreach my $record (@lines) {
			# Display the log entry:
			display_record($record);

			# Check whether the limited number of displayed entries is requested:
			if ($number > 0) {
				# Increase the displayed entries counter:
				$count++;

				# End loop when the counter reaches the limit:
				last if $count == $number;
			}
		}
	}

	# Return success:
	return 1;
}

# Set up the option parser:
Getopt::Long::Configure('no_auto_abbrev', 'no_ignore_case', 'bundling');

# Process command line options:
GetOptions(
	'help|h'               => sub { display_help();    exit 0; },
	'version|v'            => sub { display_version(); exit 0; },
	'reverse|r'            => sub { $reverse  = 1;      },
	'short|s'              => sub { $compact  = 1;      },
	'no-color|no-colour|C' => sub { $coloured = 0;      },
	'color|colour|c'       => sub { $coloured = 1;      },
	'quiet|q'              => sub { $verbose  = 0;      },
	'verbose|V'            => sub { $verbose  = 1;      },
	'blogdir|b=s'          => sub { $blogdir  = $_[1];  },
	'number|n=i'           => sub { $number   = $_[1];  },
);

# Detect superfluous options:
exit_with_error("Invalid option `$ARGV[0]'.", 22) if (scalar(@ARGV) != 0);

# Check whether the repository is present, no matter how naive this method
# actually is:
exit_with_error("Not a We-Blog repository! Try `we-blog-init' first.",1)
	unless (-d catdir($blogdir, ));

# Unless specified on the command line, read the color setup from the
# configuration file:
unless (defined $coloured) {
	# Read the configuration file:
	my $conf  = read_conf();

	# Read required data from the configuration:
	my $temp  = $conf->{color}->{log} || 'false';

	# Set up the output mode:
	$coloured = ($temp =~ /^(true|auto)\s*$/i) ? 1 : 0;
}

# Display log records:
display_log()
	or exit_with_error("Cannot read the log file.", 13);

# Return success:
exit 0;

__END__

=head1 NAME

we-blog-log - displays the We-Blog repository log

=head1 SYNOPSIS

B<we-blog-log> [B<-cqrsCV>] [B<-b> I<directory>] [B<-n> I<number>]

B<we-blog-log> B<-h>|B<-v>

=head1 DESCRIPTION

B<we-blog-log> displays the content of the We-Blog repository log.

=head1 OPTIONS

=over

=item B<-b> I<directory>, B<--blogdir> I<directory>

Allows you to specify a I<directory> in which the We-Blog repository
is placed. The default option is a current working directory.

=item B<-n> I<number>, B<--number> I<number>

Allows you to specify a I<number> of log entries to be listed.

=item B<-s>, B<--short>

Tells B<we-blog-log> to display each log entry on a single line.

=item B<-r>, B<--reverse>

Tells B<we-blog-log> to display log entries in reverse order.

=item B<-c>, B<--color>

Enables colored output. When supplied, this option overrides the relevant
configuration option.

=item B<-C>, B<--no-color>

Disables colored output. When supplied, this option overrides the relevant
configuration option.

=item B<-q>, B<--quiet>

Disables displaying of unnecessary messages.

=item B<-V>, B<--verbose>

Enables displaying of all messages. This is the default option.

=item B<-h>, B<--help>

Displays usage information and exits.

=item B<-v>, B<--version>

Displays version information and exits.

=back

=head1 FILES

=over

=item I<.we-blog/log>

A file containing the repository log.

=back

=head1 EXAMPLE USAGE

List the whole repository history:

	$ we-blog-log
	Date: Sun Jul 25 16:48:22 2010

			Edited the page with ID 5.

	Date: Tue Jul  6 18:54:59 2010

			Edited the page with ID 5.

	etc.

List the whole repository history in reverse order:

	$ we-blog-log -r
	Date: Tue Feb 10 00:40:16 2009

			Created/recovered a We-Blog repository.

	Date: Tue Feb 10 01:06:44 2009

			Added the page with ID 1.

	etc.

Display the very first log entry on a single line:

	$ we-blog-log -rs -n 1
	Tue Feb 10 00:40:16 2009 - Created/recovered a We-Blog repository.

=head1 SEE ALSO

B<we-blog-init>(1), B<we-blog-config>(1)

=head1 BUGS

To report a bug or to send a patch, please, add a new issue to the bug
tracker at <http://code.google.com/p/we-blog/issues/>, or visit the
discussion group at <https://groups.google.com/d/forum/tonk-we-blog>.

=head1 COPYRIGHT

Copyright (c) 2008-2011 Jaromir Hradilek / 2011-2012 Ton Kersten

This program is free software; see the source for copying conditions. It is
distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

=cut
